SDL/SDLJoystick.cpp

```diff
@@ -11,167 +11,209 @@ extern "C" {
+		case SDL_CONTROLLERBUTTONUP:
 		{
-			std::map<int, int>::const_iterator i = SDLJoyButtonMap.find(event.jbutton.button);
-			int deviceIndex = getDeviceIndex(event.jbutton.which);
-			if (i != SDLJoyButtonMap.end() && deviceIndex >= 0) {
-				KeyInput key;
-				key.flags = KEY_UP;
-				key.keyCode = i->second;
-				key.deviceId = DEVICE_ID_PAD_0 + deviceIndex;
-				NativeKey(key);
+			if (event.cbutton.which == player1JoyInstanceID && event.cbutton.state == SDL_RELEASED)
+			{
+				keycode_t code = getKeycodeForButton((SDL_GameControllerButton)event.cbutton.button);
+				if (code != NKCODE_UNKNOWN)
+				{
+					KeyInput key;
+					key.flags = KEY_UP;
+					key.keyCode = code;
+					key.deviceId = DEVICE_ID_PAD_0 + getDeviceIndex(event.cbutton.which);
+					NativeKey(key);
+				}
 			}
 			break;
 		}
-
-	case SDL_JOYHATMOTION:
+		case SDL_CONTROLLERAXISMOTION:
 		{
-			int deviceIndex = getDeviceIndex(event.jhat.which);
-			if (deviceIndex < 0) {
+			if (event.caxis.which == player1JoyInstanceID)
+			{
+				AxisInput axis;
+				axis.axisId = event.caxis.axis;
+				// 1.2 to try to approximate the PSP's clamped rectangular range.
+				axis.value = 1.2 * event.caxis.value / 32767.0f;
+				if (axis.value > 1.0f) axis.value = 1.0f;
+				if (axis.value < -1.0f) axis.value = -1.0f;
+				axis.deviceId = DEVICE_ID_PAD_0 + getDeviceIndex(event.caxis.which);
+				axis.flags = 0;
+				NativeAxis(axis);
 				break;
 			}
-#ifdef _WIN32
-			KeyInput key;
-			key.deviceId = DEVICE_ID_PAD_0 + deviceIndex;
-
-			key.flags = (event.jhat.value & SDL_HAT_UP)?KEY_DOWN:KEY_UP;
-			key.keyCode = NKCODE_DPAD_UP;
-			NativeKey(key);
-			key.flags = (event.jhat.value & SDL_HAT_LEFT)?KEY_DOWN:KEY_UP;
-			key.keyCode = NKCODE_DPAD_LEFT;
-			NativeKey(key);
-			key.flags = (event.jhat.value & SDL_HAT_DOWN)?KEY_DOWN:KEY_UP;
-			key.keyCode = NKCODE_DPAD_DOWN;
-			NativeKey(key);
-			key.flags = (event.jhat.value & SDL_HAT_RIGHT)?KEY_DOWN:KEY_UP;
-			key.keyCode = NKCODE_DPAD_RIGHT;
-			NativeKey(key);
-#else
-			AxisInput axisX;
-			AxisInput axisY;
-			axisX.axisId = JOYSTICK_AXIS_HAT_X;
-			axisY.axisId = JOYSTICK_AXIS_HAT_Y;
-			axisX.deviceId = DEVICE_ID_PAD_0 + deviceIndex;
-			axisY.deviceId = DEVICE_ID_PAD_0 + deviceIndex;
-			axisX.value = 0.0f;
-			axisY.value = 0.0f;
-			if (event.jhat.value & SDL_HAT_LEFT) axisX.value = -1.0f;
-			if (event.jhat.value & SDL_HAT_RIGHT) axisX.value = 1.0f;
-			if (event.jhat.value & SDL_HAT_DOWN) axisY.value = 1.0f;
-			if (event.jhat.value & SDL_HAT_UP) axisY.value = -1.0f;
-			NativeAxis(axisX);
-			NativeAxis(axisY);
-#endif
-			break;
 		}
-
-	case SDL_JOYDEVICEADDED:
+		case SDL_CONTROLLERDEVICEREMOVED:
 		{
-			int deviceIndex = event.jdevice.which;
-			if (deviceIndex >= joys.size()) {
-				joys.resize(deviceIndex+1);
+			// for removal events, "which" is the instance ID for SDL_CONTROLLERDEVICEREMOVED
+			if (event.cdevice.which == player1JoyInstanceID)
+			{
+				printf("control pad removed for player 1\n");
```

Comment:
Again, here by _player 1_ I do not mean control pad number 1... I mean the control pad being used by player 1 of the PSP. Is this ok?